terror_registry.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package errors
+
+import (
+	"fmt"
+)
+
+// Registry is a set of errors for a component.
+type Registry struct {
+	Name       string
+	errClasses map[ErrClassID]ErrClass
+}
+
+// ErrCode represents a specific error type in a error class.
+// Same error code can be used in different error classes.
+type ErrCode int
+
+// ErrCodeText is a textual error code that represents a specific error type in a error class.
+type ErrCodeText string
+
+// ErrClass represents a class of errors.
+// You can create error 'prototypes' of this class.
+type ErrClass struct {
+	ID          ErrClassID
+	Description string
+	errors      map[ErrorID]*Error
+	registry    *Registry
+}
+
+type ErrorID = string
+type ErrClassID = int
+type RFCErrorCode = string
```

Comment:
To make it clear and avoid mistakes, what about defining types?

```suggestion
type ErrorID string
type ErrClassID int
type RFCErrorCode string
```